cpp/src/sampling/random_walks_impl.cuh

```diff
@@ -133,57 +242,284 @@ struct uniform_selector {
 
       minors = std::move(sample_e_op_results);
     }
-    return std::make_tuple(std::move(minors), std::move(weights));
+    return std::make_tuple(std::move(minors), std::move(previous_vertices), std::move(weights));
   }
 };
 
 template <typename weight_t>
 struct biased_selector {
-  uint64_t seed_{0};
-
+  raft::random::RngState& rng_state_;
+  static constexpr bool is_second_order_ = false;
+  
   template <typename GraphViewType>
   std::tuple<rmm::device_uvector<typename GraphViewType::vertex_type>,
+             std::optional<rmm::device_uvector<typename GraphViewType::vertex_type>>,
              std::optional<rmm::device_uvector<weight_t>>>
   follow_random_edge(
     raft::handle_t const& handle,
     GraphViewType const& graph_view,
     std::optional<edge_property_view_t<typename GraphViewType::edge_type, weight_t const*>>
       edge_weight_view,
-    rmm::device_uvector<typename GraphViewType::vertex_type> const& current_vertices)
+    rmm::device_uvector<typename GraphViewType::vertex_type> const& current_vertices,
+    std::optional<rmm::device_uvector<typename GraphViewType::vertex_type>>& previous_vertices)
   {
     //  To do biased sampling, I need out_weights instead of out_degrees.
     //  Then I generate a random float between [0, out_weights[v]).  Then
     //  instead of making a decision based on the index I need to find
     //  upper_bound (or is it lower_bound) of the random number and
     //  the cumulative weight.
-    CUGRAPH_FAIL("biased sampling not implemented");
+
+    //  Create vertex frontier
+    using vertex_t = typename GraphViewType::vertex_type;
+
+    using tag_t = void;
+
+    cugraph::vertex_frontier_t<vertex_t, tag_t, GraphViewType::is_multi_gpu, false> vertex_frontier(handle, 1);
+
+    vertex_frontier.bucket(0).insert(current_vertices.begin(), current_vertices.end());
+
+    auto vertex_weight_sum  = compute_out_weight_sums(handle, graph_view, *edge_weight_view);
+    edge_src_property_t<GraphViewType, weight_t> edge_src_out_weight_sums(handle, graph_view);
+    update_edge_src_property(handle, graph_view, vertex_weight_sum.data(), edge_src_out_weight_sums.mutable_view());
```

Comment:
You're using source out weight sums for only the vertices in the `vertex_frontier`, right?

https://github.com/rapidsai/cugraph/blob/branch-24.10/cpp/src/prims/update_edge_src_dst_property.cuh#L938

This updates src property values for a subset of vertices. If the aggregate vertex frontier size is much smaller than the number of vertices in the graph, using this function will save both communication/computation.

You may just add a FIXME statement if you don't have time to fix it.